test/nested_derivs_unit.C

```diff
@@ -0,0 +1,37 @@
+#include "metaphysicl/dualnumberarray.h"
+
+using namespace MetaPhysicL;
+typedef double Real;
+
+typedef DualNumber<Real, NumberArray<4, Real>> ADReal;
+typedef DualNumber<ADReal, NumberArray<2, ADReal>> CompoundADType;
+
+int
+main(int argc, const char * argv[])
+{
+  ADReal u0 = 2;
+  u0.derivatives()[0] = 1;
+  ADReal u1 = 3;
+  u1.derivatives()[1] = 1;
+  ADReal v0 = 4;
+  v0.derivatives()[2] = 1;
+  ADReal v1 = 5;
+  v1.derivatives()[3] = 1;
+
+  ADReal u = 0.5 * (u0 + u1);
+  u = u + 0;
+  ADReal v = 0.5 * (v0 + v1);
+
+  const ADReal zero(0, 0);
+  const ADReal one(1, 0);
+
+  CompoundADType u_c(u, zero);
+  auto & u_cd = u_c.derivatives();
+  u_cd[0] = one;
+
+  CompoundADType v_c(v, zero);
+  auto & v_cd = v_c.derivatives();
+  v_cd[1] = one;
+
+  u_c = u_c + zero + 0;
```

Comment:
@roystgnr what I want to happen here is for `zero` to essentially get auto-promoted to a `CompoundADType` with `zero` as the `value()` and `derivatives()` corresponding to `NumberArray<2, ADReal>{}`. However, what happens without https://github.com/libMesh/MetaPhysicL/pull/36/commits/fde3fb7d47a5cb261d744e7f2981f11588147ebc is that we attempt to add the `derivatives()` of an `ADReal` with the `derivatives()` of the `CompoundADType`. This kind of thing will happen over and over for `DualNumber<T,D,asd>` and `DualNumber<T2,D2,asd>` operands. With https://github.com/libMesh/MetaPhysicL/pull/36/commits/fde3fb7d47a5cb261d744e7f2981f11588147ebc I make some progress but eventually run into the most epic block of template error message I've ever seen. A snippet:
```
../../test/navier_unit.h:119:20:   required from here
../../src/core/include/metaphysicl/compare_types.h:217:37: error: ambiguous template instantiation
for ‘struct MetaPhysicL::CompareTypes<MetaPhysicL::DualNumber<MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, double> >, MetaPhysicL::ValueLessThan>,
MetaPhysicL::ValueLessThan> >, false>,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0,
MetaPhysicL::DualNumber<double, MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0,
double>, MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<0, double> >,
MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false> >,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>, MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1,
double> >, MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false> >,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, double> >, MetaPhysicL::ValueLessThan> > > > >,
MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false>, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>, MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1,
double> >, MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false>, false, void>’
```
Is there a better way to get what I'm hoping for here? You have so many intricate unit tests in the repo; I have some belief that you've done things like this somewhere in there. With the `GeomReal` work in libMesh/MOOSE I would really like to have the same kind of type-promotion of `Real -> ADReal` for `ADReal -> CompoundADType` or else some of the programmed Newton loops in mortar, which leverage automatic differentiation to determine projections, will need more manual massaging.